rmf_fleet_adapter/src/rmf_fleet_adapter/events/internal_ReservationNodeNegotiator.hpp

```diff
@@ -190,60 +190,69 @@ class ReservationNodeNegotiator :
           }
         });
 
-
-    auto current_location = context->location();
-    if (current_location.size() == 0)
+    context->worker().schedule(
+      [ptr = negotiator->weak_from_this(), w = context->weak_from_this(), same_map](
+        const auto&)
     {
-      using namespace std::literals::chrono_literals;
-      negotiator->_retry_timer = context->node()->create_wall_timer(
-        500ms, [self = negotiator->weak_from_this(), same_map]()
-        {
-          auto negotiator = self.lock();
-          if(!negotiator)
-          {
-            return;
-          }
+      auto context = w.lock();
+      if (!context)
+        return;
 
-          auto current_location = negotiator->_context->location();
-          if (current_location.size() != 0)
-          {
-            negotiator->_retry_timer->cancel();
-          }
-          else
-          {
-            return;
-          }
-          for (std::size_t i = 0; i < negotiator->_goals.size(); ++i)
+      auto negotiator = ptr.lock();
+
+      auto current_location = context->location();
+      if (current_location.size() == 0)
+      {
+        using namespace std::literals::chrono_literals;
+        negotiator->_retry_timer = context->node()->create_wall_timer(
+          500ms, [self = negotiator->weak_from_this(), same_map]()
           {
-            if (negotiator->_goals[i].waypoint() == current_location[0].waypoint())
+            auto negotiator = self.lock();
+            if(!negotiator)
             {
-              RCLCPP_ERROR(negotiator->_context->node()->get_logger(),
-              "Already at goal no need to engage reservation system\n");
-              negotiator->_selected_final_destination_cb(negotiator->_goals[i].waypoint());
               return;
             }
+
+            auto current_location = negotiator->_context->location();
+            if (current_location.size() != 0)
+            {
+              negotiator->_retry_timer->cancel();
+            }
+            else
+            {
+              return;
+            }
+            for (std::size_t i = 0; i < negotiator->_goals.size(); ++i)
+            {
+              if (negotiator->_goals[i].waypoint() == current_location[0].waypoint())
```

Comment:
The current location is not a reliable source of truth on what point is reserved for a robot.

This should be checking if the a goal waypoint is already held by the [`context->_reservation_mgr`](https://github.com/open-rmf/rmf_ros2/blob/9e16a6e7985f97432c7a01c0ffe29b305a66b368/rmf_fleet_adapter/src/rmf_fleet_adapter/agv/RobotContext.hpp#L990).

Reply:
Thanks for spotting it. I've reworked this part and the reservation manager itself.